core/src/nodes/tasks/manager.rs

```diff
@@ -202,71 +207,52 @@ impl<I, O, H, E, HE, T, C> Manager<I, O, H, E, HE, T, C> {
         self.next_task_id.0 += 1;
 
         let (tx, rx) = mpsc::channel(4);
-        self.tasks.insert(task_id, TaskInfo { sender: tx, user_data, pending: None });
+        self.tasks.insert(task_id, TaskInfo { sender: tx, user_data });
 
-        let task: Task<futures::future::Empty<_, _>, _, _, _, _, _, _> =
+        let task: Task<Pin<Box<futures::future::Pending<_>>>, _, _, _, _, _, _> =
             Task::node(task_id, self.events_tx.clone(), rx, HandledNode::new(muxer, handler));
 
-        self.to_spawn.push(Box::new(task));
+        if let Some(threads_pool) = &mut self.threads_pool {
+            threads_pool.spawn_ok(Box::pin(task));
+        } else {
+            self.local_spawns.push(Box::pin(task));
+        }
+
         task_id
     }
 
     /// Start sending an event to all the tasks, including the pending ones.
     ///
+    /// Must be called only after a successful call to `poll_ready_broadcast`.
+    ///
     /// After starting a broadcast make sure to finish it with `complete_broadcast`,
     /// otherwise starting another broadcast or sending an event directly to a
     /// task would overwrite the pending broadcast.
     #[must_use]
-    pub fn start_broadcast(&mut self, event: &I) -> AsyncSink<()>
+    pub fn start_broadcast(&mut self, event: &I)
     where
         I: Clone
     {
-        if self.complete_broadcast().is_not_ready() {
-            return AsyncSink::NotReady(())
-        }
-
         for task in self.tasks.values_mut() {
             let msg = ToTaskMessage::HandlerEvent(event.clone());
-            task.pending = Some(AsyncSink::NotReady(msg))
+            match task.sender.start_send(msg) {
+                Ok(()) => {},
+                Err(ref err) if err.is_full() => {},        // TODO: somehow report to user?
```

Comment:
I'll add the logging, but note that this isn't supposed to happen in practice. The user is supposed to call `poll_ready_broadcast` beforehand, to make sure that room is available. And since both these functions require `&mut self`, they're not racy.

Reply:
> And since both these functions require `&mut self`, they're not racy.

I do not think this implication is valid. Given a type `T` with two operations `a` and `b`, each of which requires `&mut self`, we can wrap `T` in a `Mutex`, e.g. `Arc<Mutex<T>>` and have:

```
// thread 1 (t1: Arc<Mutex<T>>):
t1.lock().a();
t1.lock().b();

// thread 2 (t2: Arc<Mutex<T>>):
t2.lock().a();
t2.lock().b();
```

where the `MutexGuard` is dropped after each call to `a` or `b`, resulting in sequences such as `a` `a` `b` `b` for example.

Reply:
That is true for any Rust struct that exists.
The point is that you can use them in a non-racy way. We can't prevent stupidity.

Reply:
> We can't prevent stupidity.

The current API in `master` apparently can.